File: dynamicmodeling/src/main/java/org/atlanmod/model/DynamicElement.java
package org.atlanmod.model;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;

public class DynamicElement implements Serializable {

    private final String type;
    private final Map<String, Object> properties;
    private final Map<String, Object> references;

    public DynamicElement(String classname){
        this.type = classname;
        properties = new HashMap<>();
        references = new HashMap<>();
    }
    public DynamicElement(String namespace, String classname){
        this.type = namespace + ":" + classname;
        properties = new HashMap<>();
        references = new HashMap<>();
    }

    public DynamicElement(DynamicElement element){
        this.type = element.getType();
        this.properties = element.getProperties();
        this.references = element.getReferences();
    }

    public DynamicElement(DynamicElement element, Map<String, Object> properties, Map<String, Object> references){
        this.type = element.getType();
        this.properties = properties;
        this.references= references;
    }

    public  Map<String, Object> getProperties() {
        return this.properties;
    }

    public  Map<String, Object> getReferences() {
        return this.references;
    }

    public Object eGetProperty(String name){ return properties.get(name); }

    public void eSetProperty(String name, Object value){ properties.putIfAbsent(name, value); }

    public Object eGetReference(String name){ return references.get(name); }

    public void eSetReference(String name, Object value){ references.putIfAbsent(name, value); }

    public String getType(){
        return this.type;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof DynamicElement){
            DynamicElement obj_ = (DynamicElement) obj;
            return this.getType().equals(obj_.getType()) && this.getProperties().equals(obj_.getProperties());
        }
        return false;
    }

    public String toString() {
        StringBuilder str_builder = new StringBuilder();
        str_builder.append(type);
        if (properties.keySet().isEmpty()) return str_builder.toString();
        str_builder.append("\n");
        for(String property : properties.keySet()){
            str_builder.append(property).append("\n");
        }
        return str_builder.toString();
    }

}

File: dynamicmodeling/src/main/scala/org/atlanmod/model/DynamicElement.scala
package org.atlanmod.model

import scala.collection.mutable

class DynamicElement(classname: String,
                     properties : scala.collection.mutable.Map[String, Any] = mutable.HashMap())
  extends Serializable {

//    def this(element: DynamicElement, properties: scala.collection.mutable.Map[String, Any] = null) {
//        this(element.getType, if (properties == null) element.getProperties else properties)
//    }

    def getType: String = { classname }

    def getProperties: scala.collection.mutable.Map[String, Any] = { this.properties }

    def eGetProperty(name: String): Any = {
        this.properties.get(name) match {
            case Some(v) => v
            case _ => null
        }
    }

    def eSetProperty(name: String, value: Any) : Unit = {
        if (!properties.isDefinedAt(name)) properties.put(name, value)
    }

    override def equals(o : Any): Boolean = {
        o match {
            case obj: DynamicElement =>
                this.classname.equals(obj.getType) &&
                  this.properties.equals(obj.getProperties)
            case _ => false
        }
    }

    override def toString: String = {
        var res = ""
        res += classname + "\n"
        if (properties.nonEmpty) { properties.mkString("\n") }
        res
    }

}
